ntoskrnl/io/iomgr/driver.c

```diff
@@ -2181,4 +2194,237 @@ NtUnloadDriver(IN PUNICODE_STRING DriverServiceName)
+ *       Name of the service.
+ *    ServiceHandle
+ *       Handle to service registry key
+ *
+ * Return Value:
+ *    Status
+ *
+ * Status:
+ *    implemented
+ */
+
+NTSTATUS
+IopCreateLegacyDriverNode(
+    _In_ PUNICODE_STRING ServiceName,
+    _In_ HANDLE ServiceHandle)
+{
+    UNICODE_STRING RootKeyPath = RTL_CONSTANT_STRING(L"\\Registry\\Machine\\System\\CurrentControlSet\\Enum\\Root\\");
+    UNICODE_STRING LegacyPrefix = RTL_CONSTANT_STRING(L"LEGACY_");
+    UNICODE_STRING KeyString, KeyValueStr, SubKeyName, ServiceName1, DisplayName;
+    HANDLE RootKeyHandle, KeyHandle, NodeHandle;
+    PKEY_VALUE_FULL_INFORMATION KeyValueInfo;
+    ULONG RegValue;
+    NTSTATUS Status;
+
+    PAGED_CODE();
+
+    DPRINT("legacy driver: '%wZ'\n", ServiceName);
+
+    Status = IopOpenRegistryKeyEx(&RootKeyHandle,
+                                  NULL,
+                                  &RootKeyPath,
+                                  KEY_ALL_ACCESS);
+    if (!NT_SUCCESS(Status))
+        return Status;
+
+    // create sub key and try to open or create it
+    SubKeyName.MaximumLength = LegacyPrefix.Length + ServiceName->Length + sizeof(UNICODE_NULL);
+    SubKeyName.Length = 0;
+    SubKeyName.Buffer = ExAllocatePool(PagedPool, SubKeyName.MaximumLength);
+    if (SubKeyName.Buffer == NULL)
+    {
+        DPRINT1("No SubKeyName.Buffer!\n");
+        ZwClose(RootKeyHandle);
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+
+    RtlAppendUnicodeStringToString(&SubKeyName, &LegacyPrefix);
+    RtlAppendUnicodeStringToString(&SubKeyName, ServiceName);
+
+    Status = IopCreateRegistryKeyEx(&KeyHandle,
+                                    RootKeyHandle,
+                                    &SubKeyName,
+                                    KEY_ALL_ACCESS,
+                                    REG_OPTION_NON_VOLATILE,
+                                    NULL);
+
+    ZwClose(RootKeyHandle);
+    ExFreePool(SubKeyName.Buffer);
+
+    if (!NT_SUCCESS(Status))
+    {
+        ZwClose(KeyHandle);
+        return Status;
+    }
+
+    RtlInitUnicodeString(&KeyString, L"NextInstance");
+    RegValue = 1;
+    Status = ZwSetValueKey(KeyHandle,
+                           &KeyString,
+                           0,
+                           REG_DWORD,
+                           &RegValue,
+                           sizeof(RegValue));
+    if (!NT_SUCCESS(Status))
+    {
+        ZwClose(KeyHandle);
+        return Status;
+    }
+
+    RtlInitUnicodeString(&KeyString, L"000");
+    Status = IopCreateRegistryKeyEx(&NodeHandle,
+                                    KeyHandle,
+                                    &KeyString,
+                                    KEY_ALL_ACCESS,
+                                    REG_OPTION_NON_VOLATILE,
+                                    NULL);
+
+    ZwClose(KeyHandle);
+
+    if (!NT_SUCCESS(Status))
+    {
+        return Status;
+    }
+
+    ServiceName1.MaximumLength = ServiceName->Length + sizeof(UNICODE_NULL);
```

Comment:
When i'm using just the ServiceName pointer some registry string values for the Service has some trailing character(s). I was to lazy to look for the reason yet. :-) I'll have a debug session to figure out what is it about.

Reply:
That's because the `ServiceName` is a counted unicode string, that is not necessarily null-terminated. Its actual length is stored in its `Length` member.

Reply:
Well, then i would like to keep ServiceName1 and also add UNICODE_NULL at the end to create a null-terminated string.

Reply:
Probably, yes...
Then use (Status to be checked on return too):
```c
Status = RtlDuplicateUnicodeString(RTL_DUPLICATE_UNICODE_STRING_NULL_TERMINATE, ServiceName, &ServiceName1);
```
(for freeing such string, `RtlFreeUnicodeString(&ServiceName1);` ), this will be cleaner.